src/document.h

```diff
@@ -1117,12 +1115,36 @@ struct Document {
                 if (sys->frame->borddd) sys->frame->borddd->ShowPopup();
                 break;
 
+            case A_REPLACEONCE:
+            case A_REPLACEONCEJ:
             case A_REPLACEALL: {
                 if (!sys->searchstring.Len()) return _(L"No search.");
-                rootgrid->AddUndo(this);  // expensive?
-                rootgrid->FindReplaceAll(sys->frame->replaces->GetValue());
-                rootgrid->ResetChildren();
-                Refresh();
+                if (k == A_REPLACEALL) {
+                    wxString filter = sys->frame->filter->GetValue();
+                    wxString replaces = sys->frame->replaces->GetValue();
+                    loopallcells(c) if(c->text.IsInSearch()) c->AddUndo(this);
+                    rootgrid->FindReplaceAll(replaces);
+                    sys->frame->filter->SetValue(replaces);
+                    sys->frame->replaces->SetValue(filter);
+                    rootgrid->ResetChildren();
+                    Refresh();
+                } else {
+                    loopallcellssel(c, true) if(c->text.IsInSearch()) c->AddUndo(this);

```

Comment:
this seems to repeat code from the other if branch

Reply:
pull `(sys->casesensitivesearch) ? (wxString) wxEmptyString : replaces.Lower()` out of the if